code/datums/components/rot.dm

```diff
@@ -1,78 +1,133 @@
+#define TEMPERATURE_BLOCKER (1<<1)
+#define HUSK_BLOCKER (1<<2)
 
-/datum/component/rot/Initialize(new_amount)
+///Makes a thing rotting, carries with it a start delay and some things that can halt the rot, along with infection logic
+/datum/component/rot
+	///The time we were created, allows for cheese smell
+	var/start_time = 0
+	///The delay in ticks between the start of rot and effects kicking in
+	var/start_delay = 0
+	///How strong is the rot? used for scaling different aspects of the component. Between 0 and 1
+	var/strength = 0
+	///Is the component active right now?
+	var/active = FALSE
+	///Bitfield of sources preventing the component from rotting
+	var/blockers = NONE
+
+/datum/component/rot/Initialize(delay, severity)
 	if(!isatom(parent))
 		return COMPONENT_INCOMPATIBLE
+	if(isliving(parent))
+		var/mob/living/living_parent = parent
+		//I think this can break in cases where someone becomes a robot post death, but I uh, I don't know how to account for that
+		if(!(living_parent.mob_biotypes & (MOB_ORGANIC|MOB_UNDEAD)))
+			return COMPONENT_INCOMPATIBLE
+
+	start_delay = delay
+	strength = severity
+
+	RegisterSignal(parent, list(COMSIG_ATOM_HULK_ATTACK, COMSIG_ATOM_ATTACK_ANIMAL, COMSIG_ATOM_ATTACK_HAND), .proc/rot_react_touch)
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/rot_hit_react)
+	if(ismovable(parent))
+		RegisterSignal(parent, list(COMSIG_MOVABLE_CROSSED, COMSIG_MOVABLE_BUMP), .proc/rot_react)
+	if(isliving(parent))
+		RegisterSignal(parent, COMSIG_LIVING_REVIVE, .proc/react_to_revive) //mobs stop this when they come to life
+		RegisterSignal(parent, COMSIG_LIVING_GET_PULLED, .proc/rot_react_touch)
+	if(iscarbon(parent))
+		var/mob/living/carbon/carbon_parent = parent
+		RegisterSignal(carbon_parent.reagents, list(COMSIG_REAGENTS_ADD_REAGENT,
+													COMSIG_REAGENTS_REM_REAGENT,
+													COMSIG_REAGENTS_DEL_REAGENT), .proc/check_reagent)
+		RegisterSignal(parent, list(SIGNAL_ADDTRAIT(TRAIT_HUSK), SIGNAL_REMOVETRAIT(TRAIT_HUSK)), .proc/check_husk_trait)
+		check_reagent(carbon_parent.reagents, null)
+		check_husk_trait(null)
+	if(ishuman(parent))
+		var/mob/living/carbon/human/human_parent = parent
+		RegisterSignal(parent, COMSIG_HUMAN_CORETEMP_CHANGE, .proc/check_for_temperature)
+		check_for_temperature(null, 0, human_parent.coretemperature)
+
+	start_up(NONE) //If nothing's blocking it, start
+
+///One of two procs that modifies blockers, this one handles removing a blocker and potentially restarting the rot
+/datum/component/rot/proc/start_up(blocker_type)
+	blockers &= ~blocker_type //Yeet the type
+	if(blockers & ALL || active)  //If it's not empty
+		return
+	start_time = world.time
+	active = TRUE
+
+///One of two procs that modifies blockers, this one handles adding a blocker and potentially ending the rot
+/datum/component/rot/proc/rest(blocker_type)
+	var/old_blockers = blockers
+	blockers |= blocker_type
+	if(old_blockers & ALL || !active) //If it had anything before this
```

Comment:
Are you sure this is the right way to do this, and not just

```suggestion
	if(old_blockers > 0 || !active) //If it had anything before this
```

...? This would only pass if every single bit is set.

Reply:
I was trying to make the "bitfield" portion of things more clear, but you're right, this was the wrong way to go about it

Reply:
you could do `> NONE` if you really wanted to

Reply:
wouldn't get across the idea proper